interfaces/udev/backend.go

```diff
@@ -91,22 +92,46 @@ func ensureDirState(dir, glob string, content map[string]*osutil.FileState, snap
 	return errReload
 }
 
+func generateHash(bytes []byte) uint32 {
+	h := fnv.New32a()
+	h.Write(bytes)
+	return h.Sum32()
+}
+
 // combineSnippets combines security snippets collected from all the interfaces
 // affecting a given snap into a content map applicable to EnsureDirState.
 func (b *Backend) combineSnippets(snapInfo *snap.Info, snippets map[string][][]byte) (content map[string]*osutil.FileState, err error) {
+	var snapSnippets = make(map[uint32][]byte)
+	var finalSnapSnippets [][]byte
+
 	for _, appInfo := range snapInfo.Apps {
 		securityTag := appInfo.SecurityTag()
 		appSnippets := snippets[securityTag]
 		if len(appSnippets) == 0 {
 			continue
 		}
-		if content == nil {
-			content = make(map[string]*osutil.FileState)
+
+		// Add all app snippets to the snap snippet list and
+		// make sure we don't have doubles in there as they
+		// get all added to the same udev rule file in the
+		// end.
+		for _, snippet := range appSnippets {
+			snippetHash := generateHash(snippet)
+			if _, ok := snapSnippets[snippetHash]; ok {
+				continue
+			}
+			snapSnippets[snippetHash] = snippet
+			finalSnapSnippets = append(finalSnapSnippets, snippet)
 		}
+	}
 
-		addContent(securityTag, appSnippets, content)
+	if content == nil {
+		content = make(map[string]*osutil.FileState)
 	}
 
+	snapSecurityTag := snap.SecurityTag(snapInfo.Name())
+	addContent(snapSecurityTag, finalSnapSnippets, content)
+
 	for _, hookInfo := range snapInfo.Hooks {
 		securityTag := hookInfo.SecurityTag()
 		hookSnippets := snippets[securityTag]
```

Comment:
Yeah I think that makes the most sense. Apps and hooks are pretty much the same, it's just that hooks are typically run by snapd instead of directly by the user.